File: lib/domain/crop/interactors/crop_masters_task_template_index_interactor.rb
# frozen_string_literal: true

module Domain
  module Crop
    module Interactors
      class CropMastersTaskTemplateIndexInteractor
        def initialize(output_port:, gateway:, user_lookup:)
          @output_port = output_port
          @gateway = gateway
          @user_lookup = user_lookup
        end

        def call(input_dto)
          user = @user_lookup.find(input_dto.user_id)
          access_filter = Domain::Shared::Policies::CropPolicy.record_access_filter(user)
          failure = Domain::Crop::Dtos::MastersCropTaskTemplateMastersFailure.new(reason: :crop_not_found)

          begin
            crop_entity = @gateway.find_by_id(input_dto.crop_id.to_i)
          rescue Domain::Shared::Exceptions::RecordNotFound
            return @output_port.on_failure(failure)
          end

          return unless Domain::Crop::Policies::CropMastersCropEditAccess.assert_edit_or_on_failure(
            access_filter: access_filter,
            crop_entity: crop_entity,
            output_port: @output_port,
            failure: failure,
          )

          rows = @gateway.masters_crop_agricultural_task_templates_index_rows(crop_id: input_dto.crop_id)
          @output_port.on_success(rows)
        end
      end
    end
  end
end

File: crates/agrr-domain/src/crop/interactors/crop_masters_task_template_index_interactor.rs
//! Ruby: `Domain::Crop::Interactors::CropMastersTaskTemplateIndexInteractor`
use crate::crop::dtos::{MastersCropTaskTemplateIndexInput, MastersCropTaskTemplateMastersFailure, MastersCropTaskTemplateMastersFailureReason};
use crate::crop::gateways::CropGateway;
use crate::crop::policies::crop_masters_crop_edit_access;
use crate::crop::ports::CropMastersTaskTemplateIndexOutputPort;
use crate::shared::exceptions::RecordNotFoundError;
use crate::shared::gateways::UserLookupGateway;
use crate::shared::policies::crop_policy;

pub struct CropMastersTaskTemplateIndexInteractor<'a, G, O, U> { output_port: &'a mut O, gateway: &'a G, user_lookup: &'a U }
impl<'a, G, O, U> CropMastersTaskTemplateIndexInteractor<'a, G, O, U>
where G: CropGateway, O: CropMastersTaskTemplateIndexOutputPort, U: UserLookupGateway,
{
    pub fn new(output_port: &'a mut O, gateway: &'a G, user_lookup: &'a U) -> Self { Self { output_port, gateway, user_lookup } }
    pub fn call(&mut self, input: MastersCropTaskTemplateIndexInput) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let user = self.user_lookup.find(input.user_id);
        let access_filter = crop_policy::record_access_filter(user);
        let failure = MastersCropTaskTemplateMastersFailure::new(MastersCropTaskTemplateMastersFailureReason::CropNotFound);
        let crop_entity = match self.gateway.find_by_id(input.crop_id) {
            Ok(e) => e,
            Err(e) if e.downcast_ref::<RecordNotFoundError>().is_some() => { self.output_port.on_failure(failure); return Ok(()); }
            Err(e) => return Err(e),
        };
        if crop_masters_crop_edit_access::assert_edit(&access_filter, &crop_entity).is_err() {
            self.output_port.on_failure(MastersCropTaskTemplateMastersFailure::new(MastersCropTaskTemplateMastersFailureReason::CropNotFound));
            return Ok(());
        }
        let rows = self.gateway.masters_crop_agricultural_task_templates_index_rows(input.crop_id)?;
        self.output_port.on_success(rows);
        Ok(())
    }
}

#[cfg(test)]
mod interactors_crop_masters_task_template_index_interactor_test_inline {
    use super::*;
    include!(concat!(env!("CARGO_MANIFEST_DIR"), "/test/crop/interactors_crop_masters_task_template_index_interactor_test.rs"));
}
